lib/timers.js

```diff
@@ -554,67 +557,75 @@ function tryOnImmediate(immediate, queue) {
   }
 }
 
+function runCallback(timer) {
+  const argv = timer._argv;
+  const argc = argv ? argv.length : 0;
+  switch (argc) {
+    // fast-path callbacks with 0-3 arguments
+    case 0:
+      return timer._callback();
+    case 1:
+      return timer._callback(argv[0]);
+    case 2:
+      return timer._callback(argv[0], argv[1]);
+    case 3:
+      return timer._callback(argv[0], argv[1], argv[2]);
+    // more than 3 arguments run slower with .apply
+    default:
+      return timer._callback.apply(timer, argv);
+  }
+}
 
-function Immediate() { }
-
-Immediate.prototype.domain = undefined;
-Immediate.prototype._onImmediate = undefined;
-Immediate.prototype._idleNext = undefined;
-Immediate.prototype._idlePrev = undefined;
 
+function Immediate() {
+  // assigning callback here in the constructor results in pessimal performance,
+  // so have caller annotate the object (node v6.0.0, v8 5.0.71.35)
+  this._idleNext = null;
+  this._idlePrev = null;
+  this._callback = null;
+  this._argv = null;
+  this._onImmediate = null;
+  this.domain = process.domain;
+}
 
 exports.setImmediate = function(callback, arg1, arg2, arg3) {
   if (typeof callback !== 'function') {
     throw new TypeError('"callback" argument must be a function');
   }
 
   var i, args;
-  var len = arguments.length;
-  var immediate = new Immediate();
-
-  L.init(immediate);
 
-  switch (len) {
+  switch (arguments.length) {
     // fast cases
     case 0:
     case 1:
-      immediate._onImmediate = callback;
       break;
     case 2:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1);
-      };
+      args = [arg1];
       break;
     case 3:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1, arg2);
-      };
+      args = [arg1, arg2];
       break;
     case 4:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1, arg2, arg3);
-      };
+      args = [arg1, arg2, arg3];
       break;
     // slow case
     default:
-      args = new Array(len - 1);
-      for (i = 1; i < len; i++)
+      args = [arg1, arg2, arg3];
+      for (i = 4; i < arguments.length; i++)
         args[i - 1] = arguments[i];
-
-      immediate._onImmediate = function() {
-        callback.apply(immediate, args);
-      };
       break;
   }
+  var immediate = new Immediate();
```

Comment:
mscdex also suggested that, but making it `const` causes v8 to deoptimize the function.
Curiously, it's when the `new Immediate` is called below the switch that it deoptimizes,
allocating above the switch is ok.
```
timers/immediate.js thousands=2000 type=depth: ./node-var: 539.19 ./node-const: 490.39 ...... 9.95%
timers/immediate.js thousands=2000 type=depth1: ./node-var: 529.02 ./node-const: 483.8 ...... 9.35%
timers/immediate.js thousands=2000 type=breadth: ./node-var: 2378.1 ./node-const: 1568.7 ... 51.59%
timers/immediate.js thousands=2000 type=breadth1: ./node-var: 2120.1 ./node-const: 973.55 . 117.77%
timers/immediate.js thousands=2000 type=breadth4: ./node-var: 941.43 ./node-const: 668.64 .. 40.80%
timers/immediate.js thousands=2000 type=clear: ./node-var: 2247.3 ./node-const: 1004.1 .... 123.80%
```